File: typescript/src/crypto/rsa.ts
/**
 * RSA-PSS signature backend.
 *
 * Profile: RSASSA-PSS with SHA-256 digest, MGF1(SHA-256), salt length 32
 * bytes, minimum modulus 2048 bits. Keys are imported as PKCS#8 PEM (private)
 * or SPKI PEM (public); the raw-bytes material is the DER-encoded key
 * unwrapped (i.e. the bytes `openssl rsa -outform DER` emits) for the private
 * half and the inner SubjectPublicKey BIT STRING payload for the public half.
 *
 * Encrypted-at-rest PEM is NOT this module's responsibility — the key
 * hierarchy layer (`../key-hierarchy.ts`) wraps the plaintext PEM from
 * `generate()` in PBES2+AES-256-GCM before persistence.
 */

import {
  createPrivateKey,
  createPublicKey,
  constants as cryptoConstants,
  sign as cryptoSign,
  verify as cryptoVerify,
  generateKeyPairSync,
  type KeyObject,
} from 'node:crypto';

import { errors, LicensingError } from '../errors.ts';
import type {
  KeyGenerator,
  PemKeyMaterial,
  PrivateKeyHandle,
  PublicKeyHandle,
  RawKeyMaterial,
  SignatureBackend,
  Signer,
  Verifier,
} from './types.ts';

export const RSA_MIN_BITS = 2048;
export const RSA_DEFAULT_BITS = 3072;
export const RSA_PSS_SALT_LEN = 32;
export const RSA_PSS_HASH = 'sha256' as const;

interface RSAPrivateHandle extends PrivateKeyHandle {
  readonly __rsaPrivate: KeyObject;
}
interface RSAPublicHandle extends PublicKeyHandle {
  readonly __rsaPublic: KeyObject;
}

function wrapPrivate(k: KeyObject): RSAPrivateHandle {
  return { __rsaPrivate: k } as RSAPrivateHandle;
}
function wrapPublic(k: KeyObject): RSAPublicHandle {
  return { __rsaPublic: k } as RSAPublicHandle;
}
function unwrapPrivate(h: PrivateKeyHandle): KeyObject {
  return (h as RSAPrivateHandle).__rsaPrivate;
}
function unwrapPublic(h: PublicKeyHandle): KeyObject {
  return (h as RSAPublicHandle).__rsaPublic;
}

function isRaw(m: PemKeyMaterial | RawKeyMaterial): m is RawKeyMaterial {
  return 'publicRaw' in m && m.publicRaw instanceof Uint8Array;
}

function isPem(m: PemKeyMaterial | RawKeyMaterial): m is PemKeyMaterial {
  return 'publicPem' in m && typeof (m as PemKeyMaterial).publicPem === 'string';
}

/** Enforce the 2048-bit lower bound. The asymmetric key details expose a
 *  `modulusLength` field for RSA keys. */
function assertMinimumStrength(k: KeyObject): void {
  const details = k.asymmetricKeyDetails;
  if (!details || typeof details.modulusLength !== 'number') {
    // Key type wasn't RSA (or Node didn't populate details) — the alg registry
    // would already have rejected non-RSA keys at import time.
    throw errors.insufficientKeyStrength('RSA key details missing modulus length');
  }
  if (details.modulusLength < RSA_MIN_BITS) {
    throw errors.insufficientKeyStrength(
      `RSA modulus ${details.modulusLength} bits < required ${RSA_MIN_BITS}`,
    );
  }
}

class RSAPSSBackend implements SignatureBackend, Signer, Verifier, KeyGenerator {
  readonly alg = 'rs256-pss' as const;

  async importPrivate(
    material: PemKeyMaterial | RawKeyMaterial,
    passphrase?: string,
  ): Promise<PrivateKeyHandle> {
    try {
      let k: KeyObject;
      if (isRaw(material)) {
        if (!material.privateRaw) {
          throw errors.tokenMalformed('raw material has no privateRaw');
        }
        // Raw RSA private is DER-encoded PKCS#8 (or PKCS#1; we accept both).
        // Feed to Node's loader and let it figure out the type.
        k = createPrivateKey({
          key: Buffer.from(material.privateRaw),
          format: 'der',
          type: 'pkcs8',
        });
      } else if (isPem(material)) {
        if (!material.privatePem) {
          throw errors.tokenMalformed('PEM material has no privatePem');
        }
        const opts: Parameters<typeof createPrivateKey>[0] = { key: material.privatePem };
        if (passphrase !== undefined) {
          (opts as { passphrase?: string }).passphrase = passphrase;
        }
        k = createPrivateKey(opts);
      } else {
        throw errors.tokenMalformed('unrecognized key material shape');
      }
      if (k.asymmetricKeyType !== 'rsa' && k.asymmetricKeyType !== 'rsa-pss') {
        throw errors.insufficientKeyStrength(
          `expected RSA key, got ${String(k.asymmetricKeyType)}`,
        );
      }
      assertMinimumStrength(k);
      return wrapPrivate(k);
    } catch (e) {
      if (e instanceof LicensingError) throw e;
      throw errors.keyDecryptionFailed();
    }
  }

  async importPublic(material: PemKeyMaterial | RawKeyMaterial): Promise<PublicKeyHandle> {
    let k: KeyObject;
    if (isRaw(material)) {
      k = createPublicKey({ key: Buffer.from(material.publicRaw), format: 'der', type: 'spki' });
    } else if (isPem(material)) {
      k = createPublicKey(material.publicPem);
    } else {
      throw errors.tokenMalformed('unrecognized key material shape');
    }
    if (k.asymmetricKeyType !== 'rsa' && k.asymmetricKeyType !== 'rsa-pss') {
      throw errors.insufficientKeyStrength(`expected RSA key, got ${String(k.asymmetricKeyType)}`);
    }
    assertMinimumStrength(k);
    return wrapPublic(k);
  }

  async sign(key: PrivateKeyHandle, data: Uint8Array): Promise<Uint8Array> {
    const sig = cryptoSign(RSA_PSS_HASH, data, {
      key: unwrapPrivate(key),
      padding: cryptoConstants.RSA_PKCS1_PSS_PADDING,
      saltLength: RSA_PSS_SALT_LEN,
    });
    return new Uint8Array(sig.buffer, sig.byteOffset, sig.byteLength);
  }

  async verify(key: PublicKeyHandle, data: Uint8Array, signature: Uint8Array): Promise<boolean> {
    return cryptoVerify(
      RSA_PSS_HASH,
      data,
      {
        key: unwrapPublic(key),
        padding: cryptoConstants.RSA_PKCS1_PSS_PADDING,
        saltLength: RSA_PSS_SALT_LEN,
      },
      signature,
    );
  }

  async generate(_passphrase: string): Promise<{ pem: PemKeyMaterial; raw: RawKeyMaterial }> {
    // 3072-bit default (comfortably above the 2048 floor; matches fixture).
    const { privateKey, publicKey } = generateKeyPairSync('rsa', {
      modulusLength: RSA_DEFAULT_BITS,
      publicExponent: 0x10001,
    });
    const privatePem = privateKey.export({ format: 'pem', type: 'pkcs8' }).toString();
    const publicPem = publicKey.export({ format: 'pem', type: 'spki' }).toString();
    const privDer = privateKey.export({ format: 'der', type: 'pkcs8' });
    const pubDer = publicKey.export({ format: 'der', type: 'spki' });
    return {
      pem: { privatePem, publicPem },
      raw: {
        privateRaw: new Uint8Array(privDer.buffer, privDer.byteOffset, privDer.byteLength),
        publicRaw: new Uint8Array(pubDer.buffer, pubDer.byteOffset, pubDer.byteLength),
      },
    };
  }
}

export const rsaPssBackend: SignatureBackend = new RSAPSSBackend();

File: licensing/crypto/rsa/rsa.go
// Package rsa implements the licensing SignatureBackend for RSASSA-PSS with
// SHA-256 digest, MGF1(SHA-256), and salt length 32 bytes (equal to the hash
// output length).
//
// Uses only the Go standard library (crypto/rsa, crypto/x509, encoding/pem).
// Mirrors @anorebel/licensing/crypto/rsa semantically:
//
//   - Accepts PKCS#8 PEM (plaintext) or raw DER (PKCS#8) for private keys
//   - Accepts SPKI PEM or raw SPKI DER for public keys
//   - Enforces a 2048-bit modulus floor; default generate size is 3072
//   - Deterministic on the verify side (PSS sig length == modulus length)
//
// Encrypted-at-rest PEM wrapping is the key-hierarchy layer's responsibility,
// not this backend's.
package rsa

import (
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/pem"
	"fmt"

	lic "github.com/AnoRebel/licensing/licensing"
)

// Profile parameters — these MUST match @anorebel/licensing/crypto/rsa.ts exactly
// or cross-language verification will fail silently.
const (
	MinBits     = 2048
	DefaultBits = 3072
	PSSSaltLen  = 32 // = SHA-256 output length; matches TS `saltLength: 32`
)

var pssOptions = &rsa.PSSOptions{
	SaltLength: PSSSaltLen,
	Hash:       crypto.SHA256,
}

// Backend is the sole SignatureBackend for RSA-PSS. Stateless.
type Backend struct{}

// New returns a ready-to-register backend value.
func New() Backend { return Backend{} }

// Alg returns the algorithm identifier this backend handles.
func (Backend) Alg() lic.KeyAlg { return lic.AlgRSAPSS }

// privateHandle wraps an *rsa.PrivateKey plus the PKCS#8 DER bytes so callers
// that want the raw material back don't have to re-marshal.
type privateHandle struct {
	key *rsa.PrivateKey
	der []byte // PKCS#8 DER
}

// publicHandle wraps an *rsa.PublicKey plus its SPKI DER.
type publicHandle struct {
	key *rsa.PublicKey
	der []byte // SPKI DER
}

// ImportPrivate parses PKCS#8 PEM or raw PKCS#8 DER into an opaque handle.
// Raw takes priority when both fields are populated. `passphrase` is unused
// here — encrypted PKCS#8 is unwrapped by the key-hierarchy layer before
// reaching this backend.
func (Backend) ImportPrivate(m lic.KeyMaterial, _ string) (lic.PrivateKeyHandle, error) {
	// Raw DER path.
	if len(m.Raw.PrivateRaw) > 0 {
		return importPrivateDER(m.Raw.PrivateRaw)
	}
	// PEM path.
	if m.Pem.PrivatePem != "" {
		blk, _ := pem.Decode([]byte(m.Pem.PrivatePem))
		if blk == nil {
			return nil, wrapMalformed("rsa: invalid PEM envelope")
		}
		if blk.Type != "PRIVATE KEY" {
			return nil, wrapMalformed(fmt.Sprintf(
				"rsa: expected PEM type \"PRIVATE KEY\", got %q", blk.Type))
		}
		return importPrivateDER(blk.Bytes)
	}
	return nil, wrapMalformed("rsa: neither Raw.PrivateRaw nor Pem.PrivatePem provided")
}

func importPrivateDER(der []byte) (lic.PrivateKeyHandle, error) {
	keyAny, err := x509.ParsePKCS8PrivateKey(der)
	if err != nil {
		return nil, wrapDecryptFailed()
	}
	priv, ok := keyAny.(*rsa.PrivateKey)
	if !ok {
		return nil, wrapStrength(fmt.Sprintf(
			"rsa: PKCS#8 key is %T, not *rsa.PrivateKey", keyAny))
	}
	if priv.N.BitLen() < MinBits {
		return nil, wrapStrength(fmt.Sprintf(
			"rsa: modulus %d bits < required %d", priv.N.BitLen(), MinBits))
	}
	derCopy := make([]byte, len(der))
	copy(derCopy, der)
	return &privateHandle{key: priv, der: derCopy}, nil
}

// ImportPublic parses SPKI PEM or raw SPKI DER.
func (Backend) ImportPublic(m lic.KeyMaterial) (lic.PublicKeyHandle, error) {
	if len(m.Raw.PublicRaw) > 0 {
		return importPublicDER(m.Raw.PublicRaw)
	}
	if m.Pem.PublicPem != "" {
		blk, _ := pem.Decode([]byte(m.Pem.PublicPem))
		if blk == nil {
			return nil, wrapMalformed("rsa: invalid PEM envelope")
		}
		if blk.Type != "PUBLIC KEY" {
			return nil, wrapMalformed(fmt.Sprintf(
				"rsa: expected PEM type \"PUBLIC KEY\", got %q", blk.Type))
		}
		return importPublicDER(blk.Bytes)
	}
	return nil, wrapMalformed("rsa: neither Raw.PublicRaw nor Pem.PublicPem provided")
}

func importPublicDER(der []byte) (lic.PublicKeyHandle, error) {
	keyAny, err := x509.ParsePKIXPublicKey(der)
	if err != nil {
		return nil, wrapMalformed("rsa: SPKI parse failed: " + err.Error())
	}
	pub, ok := keyAny.(*rsa.PublicKey)
	if !ok {
		return nil, wrapStrength(fmt.Sprintf(
			"rsa: SPKI key is %T, not *rsa.PublicKey", keyAny))
	}
	if pub.N.BitLen() < MinBits {
		return nil, wrapStrength(fmt.Sprintf(
			"rsa: modulus %d bits < required %d", pub.N.BitLen(), MinBits))
	}
	derCopy := make([]byte, len(der))
	copy(derCopy, der)
	return &publicHandle{key: pub, der: derCopy}, nil
}

// Sign returns an RSASSA-PSS signature over data. The input is hashed with
// SHA-256 before signing, matching Node's `sign('sha256', data, …)`.
func (Backend) Sign(h lic.PrivateKeyHandle, data []byte) ([]byte, error) {
	ph, ok := h.(*privateHandle)
	if !ok {
		return nil, wrapMalformed(fmt.Sprintf(
			"rsa.Sign: wrong handle type %T", h))
	}
	digest := sha256.Sum256(data)
	sig, err := rsa.SignPSS(rand.Reader, ph.key, crypto.SHA256, digest[:], pssOptions)
	if err != nil {
		return nil, wrapSignatureInvalid()
	}
	return sig, nil
}

// Verify returns (true, nil) iff sig is a valid RSA-PSS signature over data
// under the key in h. A structurally bad signature (wrong length, bad
// padding, bad digest) returns (false, nil) — matching the TS contract that
// surfaces ErrTokenSignatureInvalid at the Verify() layer, not here.
func (Backend) Verify(h lic.PublicKeyHandle, data, sig []byte) (bool, error) {
	ph, ok := h.(*publicHandle)
	if !ok {
		return false, wrapMalformed(fmt.Sprintf(
			"rsa.Verify: wrong handle type %T", h))
	}
	// Cheap pre-check: PSS signature length always equals the modulus size
	// in bytes. Save a full verify op on obviously malformed input.
	if len(sig) != (ph.key.N.BitLen()+7)/8 {
		return false, nil
	}
	digest := sha256.Sum256(data)
	if err := rsa.VerifyPSS(ph.key, crypto.SHA256, digest[:], sig, pssOptions); err != nil {
		return false, nil
	}
	return true, nil
}

// Generate produces a fresh 3072-bit RSA keypair with plaintext PKCS#8 PEM
// plus raw DER for both halves. `passphrase` is ignored here; encryption is
// a separate layer.
func (Backend) Generate(_ string) (lic.PemKeyMaterial, lic.RawKeyMaterial, error) {
	priv, err := rsa.GenerateKey(rand.Reader, DefaultBits)
	if err != nil {
		return lic.PemKeyMaterial{}, lic.RawKeyMaterial{}, err
	}
	pkcs8, err := x509.MarshalPKCS8PrivateKey(priv)
	if err != nil {
		return lic.PemKeyMaterial{}, lic.RawKeyMaterial{}, err
	}
	spki, err := x509.MarshalPKIXPublicKey(&priv.PublicKey)
	if err != nil {
		return lic.PemKeyMaterial{}, lic.RawKeyMaterial{}, err
	}
	privPem := string(pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: pkcs8}))
	pubPem := string(pem.EncodeToMemory(&pem.Block{Type: "PUBLIC KEY", Bytes: spki}))

	privDer := make([]byte, len(pkcs8))
	copy(privDer, pkcs8)
	pubDer := make([]byte, len(spki))
	copy(pubDer, spki)

	return lic.PemKeyMaterial{PrivatePem: privPem, PublicPem: pubPem},
		lic.RawKeyMaterial{PrivateRaw: privDer, PublicRaw: pubDer},
		nil
}

// -----------------------------------------------------------------------
// Raw-bytes accessors. Expose the DER bytes held by each handle, copied
// so callers can't mutate backend-internal state.
// -----------------------------------------------------------------------

// PrivateDER returns a copy of the PKCS#8 DER for a handle produced by
// ImportPrivate or Generate.
func PrivateDER(h lic.PrivateKeyHandle) ([]byte, error) {
	ph, ok := h.(*privateHandle)
	if !ok {
		return nil, wrapMalformed(fmt.Sprintf(
			"rsa.PrivateDER: wrong handle type %T", h))
	}
	out := make([]byte, len(ph.der))
	copy(out, ph.der)
	return out, nil
}

// PublicDER returns a copy of the SPKI DER for a public handle.
func PublicDER(h lic.PublicKeyHandle) ([]byte, error) {
	ph, ok := h.(*publicHandle)
	if !ok {
		return nil, wrapMalformed(fmt.Sprintf(
			"rsa.PublicDER: wrong handle type %T", h))
	}
	out := make([]byte, len(ph.der))
	copy(out, ph.der)
	return out, nil
}

// Compile-time assertion.
var _ lic.SignatureBackend = Backend{}

// -----------------------------------------------------------------------
// Error wrappers — same pattern as the ed25519 backend.
// -----------------------------------------------------------------------

func wrapStrength(msg string) error {
	return fmt.Errorf("%w: %s", lic.ErrInsufficientKeyStrength, msg)
}

func wrapMalformed(msg string) error {
	return fmt.Errorf("%w: %s", lic.ErrTokenMalformed, msg)
}

func wrapDecryptFailed() error {
	return fmt.Errorf("%w", lic.ErrKeyDecryptionFailed)
}

func wrapSignatureInvalid() error {
	return fmt.Errorf("%w", lic.ErrTokenSignatureInvalid)
}
